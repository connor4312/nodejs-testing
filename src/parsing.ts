import type { Options } from "acorn";
import { parse } from "acorn-loose";
import * as evk from "eslint-visitor-keys";
import {
  CallExpression,
  Expression,
  Node,
  SourceLocation,
  Super,
  type ImportDeclaration,
} from "estree";
import * as Path from "node:path";
import * as vscode from "vscode";
import {
  defaultTestFunctionSpecifiers,
  type TestFunctionSpecifierConfig,
} from "./test-function-specifier-config";
import { assertUnreachable } from "./utils";

const enum C {
  ImportDeclaration = "ImportDeclaration",
  ImportNamespaceSpecifier = "ImportNamespaceSpecifier",
  ImportDefaultSpecifier = "ImportDefaultSpecifier",
  ImportSpecifier = "ImportSpecifier",
  MemberExpression = "MemberExpression",
  CallExpression = "CallExpression",
  VariableDeclarator = "VariableDeclarator",
  ObjectPattern = "ObjectPattern",
  ParenthesizedExpression = "ParenthesizedExpression",
  SequenceExpression = "SequenceExpression",
  TemplateLiteral = "TemplateLiteral",
  Property = "Property",
  Literal = "Literal",
  Identifier = "Identifier",
  NodeTest = "node:test",
}

export const acornOptions: Options = {
  ecmaVersion: "latest",
  locations: true,
  allowAwaitOutsideFunction: true,
  allowImportExportEverywhere: true,
  allowReserved: true,
  allowReturnOutsideFunction: true,
};

type ExtractTest = (node: CallExpression) => string | undefined;

const unpackCalleeExpression = (n: CallExpression): Expression | Super =>
  n.callee.type === C.SequenceExpression
    ? n.callee.expressions[n.callee.expressions.length - 1]
    : n.callee;

const matchIdentified =
  (name: string, alias: string = name): ExtractTest =>
  (n) => {
    const callee = unpackCalleeExpression(n);
    return callee.type === C.Identifier && callee.name === alias ? name : undefined;
  };

const matchNamespaced =
  (name: string): ExtractTest =>
  (n) => {
    const callee = unpackCalleeExpression(n);
    if (callee.type === C.Identifier && callee.name === name) {
      return "test"; // default export, #42
    }

    return callee.type === C.MemberExpression &&
      callee.object.type === C.Identifier &&
      callee.object.name === name &&
      callee.property.type === C.Identifier
      ? callee.property.name
      : undefined;
  };

const getStringish = (nameArg: Node | undefined): string | undefined => {
  if (nameArg?.type === C.Literal && typeof nameArg.value === "string") {
    return nameArg.value;
  }
  if (nameArg?.type === C.TemplateLiteral && nameArg.quasis.length === 1) {
    return nameArg.quasis[0].value.cooked || nameArg.quasis[0].value.raw;
  }
};

export interface IParsedNode {
  fn: string;
  name: string;
  location: SourceLocation;
  children: IParsedNode[];
}

/**
 * Look for test function imports in this AST node
 *
 * @param folder The workspace folder this file belongs to, used for relative path references to a custom test function
 * @param fileUri The URI of the file we are extracting from
 * @param testFunctions the tests function imports to check for
 * @param node the ImportDelcaration to look for test imports
 * @returns
 */
function importDeclarationExtractTests(
  folder: vscode.WorkspaceFolder | undefined,
  fileUri: vscode.Uri | undefined,
  testFunctions: TestFunctionSpecifierConfig[],
  node: ImportDeclaration,
): ExtractTest[] {
  const idTests: ExtractTest[] = [];
  if (typeof node.source.value !== "string") {
    return [];
  }

  let importValue = node.source.value;
  if (node.source.value.startsWith("./") || node.source.value.startsWith("../")) {
    if (!folder || !fileUri) {
      console.warn(`Trying to match custom test function without specifying a folder or fileUri`);
      return [];
    }

    // This is a relative import, we need to adjust the import value for matching purposes
    const importRelativeToRoot = Path.relative(
      folder.uri.fsPath,
      Path.resolve(Path.dirname(fileUri.fsPath), node.source.value),
    );

    importValue = `./${importRelativeToRoot}`;
  }

  for (const specifier of testFunctions) {
    if (specifier.import !== importValue) {
      continue;
    }

    // Next check to see if the functions imported are tests functions
    const validNames = new Set(
      typeof specifier.name === "string" ? [specifier.name] : specifier.name,
    );

    for (const spec of node.specifiers) {
      const specType = spec.type;
      if (specType === C.ImportDefaultSpecifier || specType === C.ImportNamespaceSpecifier) {
        if (validNames.has("default")) {
          idTests.push(matchNamespaced(spec.local.name));
        }
      } else if (specType === C.ImportSpecifier) {
        if (spec.imported.type === C.Identifier) {
          if (validNames.has(spec.imported.name)) {
            idTests.push(matchIdentified(spec.imported.name, spec.local.name));
          }
        }
      } else {
        assertUnreachable(specType, `${specType} was unhandled`);
      }
    }
  }

  return idTests;
}

export const parseSource = (
  text: string,
  folder?: vscode.WorkspaceFolder,
  fileUri?: vscode.Uri,
  testFunctions?: TestFunctionSpecifierConfig[],
): IParsedNode[] => {
  const ast = parse(text, acornOptions);
  const testMatchers = testFunctions ?? defaultTestFunctionSpecifiers;

  const idTests: ExtractTest[] = [];

  // Since tests can be nested inside of each other, for example a test suite.
  // We keep track of the test declarations in a tree.
  const stack: { node: Node; r: IParsedNode }[] = [];
  stack.push({ node: undefined, r: { children: [] } } as any);

  traverse(ast as Node, {
    enter(node) {
      if (node.type === C.ImportDeclaration) {
        const matchers = importDeclarationExtractTests(folder, fileUri, testMatchers, node);
        idTests.push(...matchers);
      } else if (
        node.type === C.VariableDeclarator &&
        node.init?.type === C.CallExpression &&
        node.init.callee.type === C.Identifier &&
        node.init.callee.name === "require"
      ) {
        const firstArg = getStringish(node.init.arguments[0]);
        if (firstArg === C.NodeTest) {
          if (node.id.type === C.ObjectPattern) {
            for (const prop of node.id.properties) {
              if (
                prop.type === C.Property &&
                prop.key.type === C.Identifier &&
                prop.value.type === C.Identifier
              ) {
                idTests.push(matchIdentified(prop.key.name, prop.value.name));
              }
            }
          } else if (node.id.type === C.Identifier) {
            idTests.push(matchNamespaced(node.id.name));
          }
        }
      } else if (node.type === C.CallExpression) {
        const name = getStringish(node.arguments[0]);
        if (name === undefined) {
          return;
        }

        for (const test of idTests) {
          const fn = test(node);
          if (fn) {
            const child: IParsedNode = {
              children: [],
              location: node.loc!,
              fn,
              name,
            };

            // We have encountered a test function, record it in the tree.
            stack[stack.length - 1].r.children.push(child);

            // This test function is potentially a "parent" for subtests, so
            // keep it as the "current leaf" of the stack, so future sub-tests
            // can be associated with it
            stack.push({ node, r: child });
            break;
          }
        }
      }
    },
    leave(node) {
      // We are exiting a node that was potentially a test function.  If it was,
      // we need to pop it of the stack, since there are no more subtests to be
      // associated with it.
      if (stack[stack.length - 1].node === node) {
        stack.pop();
      }
    },
  });

  return stack[0].r.children;
};

const traverse = (
  node: Node | undefined,
  visitor: {
    enter: (node: Node, parent?: Node) => void;
    leave: (node: Node) => void;
  },
  parent?: Node,
) => {
  if (!node) {
    return;
  }

  visitor.enter(node, parent);

  const keys = evk.KEYS[node.type];
  if (keys) {
    for (const key of keys) {
      const child = (node as unknown as Record<string, Node | Node[]>)[key];
      if (child instanceof Array) {
        for (const [i, c] of child.entries()) {
          traverse(c, visitor, node);
        }
      } else if (child) {
        traverse(child, visitor, node);
      }
    }
  }

  visitor.leave(node);
};
