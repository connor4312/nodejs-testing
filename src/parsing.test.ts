import { describe, expect, it } from "vitest";
import { parseSource } from "./parsing";
import {
  defaultTestFunctionSpecifiers,
  type TestFunctionSpecifierConfig,
} from "./test-function-specifier-config";

function parseSourceSimple(contents: string) {
  return parseSource(
    contents,
    "/workspace/",
    "/workspace/test.test.js",
    defaultTestFunctionSpecifiers,
  );
}

type ParseCustomOptions = {
  testNames?: string[];
  testImport?: string;
  workspace?: string;
  path?: string;
};

// parseSourceCustom is a ergonomic version of parseSource that has reasonable defaults for the parameters
const parseSourceCustom = (contents: string, options?: ParseCustomOptions) => {
  const testNames = options?.testNames ?? ["test"];
  const testImport = options?.testImport ?? "./test/utils";
  const workspace = options?.workspace ?? "/workspace/";
  const path = options?.path ?? "/workspace/test/addition.test.js";
  const testFunctions: TestFunctionSpecifierConfig[] = [{ import: testImport, name: testNames }];
  return parseSource(contents, workspace, path, testFunctions);
};

const testCases = (prefix = "") => `${prefix}describe("math", () => {
  ${prefix}it("addition", () => {
    strictEqual(1 + 1, 2);
  });
  ${prefix}it("addition", () => {
    strictEqual(1 + 1, 2);
  });

  ${prefix}it("subtraction", () => {
    strictEqual(1 - 1, 0);
  });
});`;

describe("extract", () => {
  it("extracts default import", () => {
    const src = `import nt from "node:test";

    ${testCases("nt.")}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("extracts star import", () => {
    const src = `import * as nt from "node:test";

    ${testCases("nt.")}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("extracts named imports", () => {
    const src = `import { describe, it } from "node:test";

    ${testCases()}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("extracts aliased imports", () => {
    const src = `import { describe as xdescribe, it as xit } from "node:test";

    ${testCases("x")}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("extracts default require", () => {
    const src = `const nt = require("node:test");

    ${testCases("nt.")}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("extracts destructed require", () => {
    const src = `const { describe, it } = require("node:test");

    ${testCases()}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("extracts aliased require", () => {
    const src = `const { describe: xdescribe, it: xit } = require("node:test");

    ${testCases("x")}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("extracts ts import mangled", () => {
    const src = `"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("assert");
const node_test_1 = require("node:test");
(0, node_test_1.describe)("math", () => {
    (0, node_test_1.it)("addition", () => {
        (0, assert_1.strictEqual)(1 + 1, 2);
    });
    (0, node_test_1.it)("addition", () => {
        (0, assert_1.strictEqual)(1 + 1, 2);
    });
    (0, node_test_1.it)("subtraction", () => {
        (0, assert_1.strictEqual)(1 - 1, 0);
    });
});
//# sourceMappingURL=example.js.map`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("does not break on empty call expression (#3)", () => {
    const src = `const { describe, it } = require("node:test");

    (function() {})();

    ${testCases()}`;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("works with string literals", () => {
    const src = `
      const nt = require(\`node:test\`);

      nt.describe(\`math\`, () => {
        nt.it(\`addition\`, () => {});
      });
    `;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("works with default cjs import", () => {
    const src = `
      const nt = require(\`node:test\`);
      nt(\`addition\`, () => {});
    `;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });

  it("works with default esm import", () => {
    const src = `
      import nt from "node:test";
      nt(\`addition\`, () => {});
    `;

    expect(parseSourceSimple(src)).toMatchSnapshot();
  });
});

describe("extract with custom test specifiers in ESM code", () => {
  it("extracts default import tests", () => {
    const contents = `
import nt from "./utils";

nt("default import test", () => {
  strictEqual(1 + 1, 2);
  nt("nested test", () => {
    strictEqual(1 + 1, 2);
  });
});
`;

    const result = parseSourceCustom(contents, {
      testNames: ["default"],
    });

    // One test, and one subtest
    expect(result.length).toEqual(1);
    expect(result[0].children.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  it("extracts star import tests", () => {
    const contents = `import  * as Utils from "./utils";
    Utils.test("addition", () => {
      strictEqual(1 + 1, 2);

      Utils.test("subtest", () => {
        strictEqual(1 + 1, 2);
      });
    });`;

    const result = parseSourceCustom(contents, {
      testNames: ["test"],
    });

    expect(result.length).toEqual(1);
    expect(result[0].children.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  it("extracts named import tests", () => {
    const contents = `import { wrappedTest } from "./utils";
    wrappedTest("addition", () => {
      strictEqual(1 + 1, 2);
    });`;

    const result = parseSourceCustom(contents, {
      testNames: ["wrappedTest"],
    });
    expect(result.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  it("extracts renamed named import tests", () => {
    const contents = `import { wrappedTest as renamedTest } from "./utils";
    renamedTest("addition", () => {
      strictEqual(1 + 1, 2);
    });`;

    const result = parseSourceCustom(contents, {
      testNames: ["wrappedTest"],
    });
    expect(result.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  it("extracts renamed named import test", () => {
    const contents = `import { wrappedTest as renamedTest } from "./utils";
    renamedTest("addition", () => {
      strictEqual(1 + 1, 2);
    });`;

    const result = parseSourceCustom(contents, {
      testNames: ["wrappedTest"],
    });
    expect(result.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });
});

describe("extract with custom test specifiers in commonjs code", () => {
  it("extracts default require tests", () => {
    const contents = `const nt = require("./utils");

nt("default import test", () => {
  strictEqual(1 + 1, 2);
  nt("nested test", () => {
    strictEqual(1 + 1, 2);
  });
});
`;
    const result = parseSourceCustom(contents, {
      testNames: ["default"],
    });

    // One test, and one subtest
    expect(result.length).toEqual(1);
    expect(result[0].children.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  it("extracts destructed require", () => {
    const contents = `const { describe, it } = require("./utils");

describe("destructured test", () => {
  strictEqual(1 + 1, 2);
  it("nested test", () => {
    strictEqual(1 + 1, 2);
  });
});
`;
    const result = parseSourceCustom(contents, {
      testNames: ["describe", "it"],
    });

    // One test, and one subtest
    expect(result.length).toEqual(1);
    expect(result[0].children.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  it("extracts aliased require", () => {
    const contents = `const { describe: xdescribe, it: xit } = require("./utils");

xdescribe("destructured test", () => {
  strictEqual(1 + 1, 2);
  xit("nested test", () => {
    strictEqual(1 + 1, 2);
  });
});
`;

    const result = parseSourceCustom(contents, {
      testNames: ["describe", "it"],
    });

    // One test, and one subtest
    expect(result.length).toEqual(1);
    expect(result[0].children.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  it("extracts ts import mangled", () => {
    const contents = `"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("assert");
const node_test_1 = require("./utils");
(0, node_test_1.describe)("math", () => {
    (0, node_test_1.it)("addition", () => {
        (0, assert_1.strictEqual)(1 + 1, 2);
    });
    (0, node_test_1.it)("addition", () => {
        (0, assert_1.strictEqual)(1 + 1, 2);
    });
    (0, node_test_1.it)("subtraction", () => {
        (0, assert_1.strictEqual)(1 - 1, 0);
    });
});
//# sourceMappingURL=example.js.map`;

    const result = parseSourceCustom(contents, {
      testNames: ["describe", "it"],
    });

    // 1 test, 3 subtests
    expect(result.length).toEqual(1);
    expect(result[0].children.length).toEqual(3);
    expect(result).toMatchSnapshot();
  });
});
