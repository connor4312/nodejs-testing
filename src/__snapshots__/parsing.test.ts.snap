// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract > does not break on empty call expression (#3) 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 8,
          },
          "start": Position {
            "column": 2,
            "line": 6,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 11,
          },
          "start": Position {
            "column": 2,
            "line": 9,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 15,
          },
          "start": Position {
            "column": 2,
            "line": 13,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 16,
      },
      "start": Position {
        "column": 4,
        "line": 5,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts aliased imports 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 6,
          },
          "start": Position {
            "column": 2,
            "line": 4,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 9,
          },
          "start": Position {
            "column": 2,
            "line": 7,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 13,
          },
          "start": Position {
            "column": 2,
            "line": 11,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 14,
      },
      "start": Position {
        "column": 4,
        "line": 3,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts aliased require 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 6,
          },
          "start": Position {
            "column": 2,
            "line": 4,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 9,
          },
          "start": Position {
            "column": 2,
            "line": 7,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 13,
          },
          "start": Position {
            "column": 2,
            "line": 11,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 14,
      },
      "start": Position {
        "column": 4,
        "line": 3,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts default import 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 6,
          },
          "start": Position {
            "column": 2,
            "line": 4,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 9,
          },
          "start": Position {
            "column": 2,
            "line": 7,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 13,
          },
          "start": Position {
            "column": 2,
            "line": 11,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 14,
      },
      "start": Position {
        "column": 4,
        "line": 3,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts default require 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 6,
          },
          "start": Position {
            "column": 2,
            "line": 4,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 9,
          },
          "start": Position {
            "column": 2,
            "line": 7,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 13,
          },
          "start": Position {
            "column": 2,
            "line": 11,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 14,
      },
      "start": Position {
        "column": 4,
        "line": 3,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts destructed require 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 6,
          },
          "start": Position {
            "column": 2,
            "line": 4,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 9,
          },
          "start": Position {
            "column": 2,
            "line": 7,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 13,
          },
          "start": Position {
            "column": 2,
            "line": 11,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 14,
      },
      "start": Position {
        "column": 4,
        "line": 3,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts named imports 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 6,
          },
          "start": Position {
            "column": 2,
            "line": 4,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 9,
          },
          "start": Position {
            "column": 2,
            "line": 7,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 13,
          },
          "start": Position {
            "column": 2,
            "line": 11,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 14,
      },
      "start": Position {
        "column": 4,
        "line": 3,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts star import 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 6,
          },
          "start": Position {
            "column": 2,
            "line": 4,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 9,
          },
          "start": Position {
            "column": 2,
            "line": 7,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 4,
            "line": 13,
          },
          "start": Position {
            "column": 2,
            "line": 11,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 14,
      },
      "start": Position {
        "column": 4,
        "line": 3,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > extracts ts import mangled 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 8,
          },
          "start": Position {
            "column": 4,
            "line": 6,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 11,
          },
          "start": Position {
            "column": 4,
            "line": 9,
          },
        },
        "name": "addition",
      },
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 6,
            "line": 14,
          },
          "start": Position {
            "column": 4,
            "line": 12,
          },
        },
        "name": "subtraction",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 2,
        "line": 15,
      },
      "start": Position {
        "column": 0,
        "line": 5,
      },
    },
    "name": "math",
  },
]
`;

exports[`extract > works with string literals 1`] = `
[
  {
    "children": [
      {
        "children": [],
        "fn": "it",
        "location": SourceLocation {
          "end": Position {
            "column": 35,
            "line": 5,
          },
          "start": Position {
            "column": 8,
            "line": 5,
          },
        },
        "name": "addition",
      },
    ],
    "fn": "describe",
    "location": SourceLocation {
      "end": Position {
        "column": 8,
        "line": 6,
      },
      "start": Position {
        "column": 6,
        "line": 4,
      },
    },
    "name": "math",
  },
]
`;
